addons/udes_stock/models/stock_move.py

```diff
@@ -143,52 +143,102 @@ def split_out_move_lines(self, move_lines, **kwargs):
+        """
+        Split the move in self.
+        If mls is None: Retain all mls with qty_done == product_uom_qty in self
+        and move all other work into the new move.
+        If mls is not None: reatian mls_to_keep in self, and remove all other work,
+        into the new move. 
+        
+
+        :kwargs:
+            - mls_to_keep: The move lines to keep in self, if None use all mls with
+            qty_done == product_uom_qty.
+        :returns: 
+            - If a move has nothing complete, return self, unattached from the pickng
+            - If a move has everything complete, return an empty record set
+            - If the move is partially complete, return the new move with the incomplete
+            work. 
+        :raises: An error if any move lines attached to self are partially complete.
+        :raises: An error if any move lines being moved are done, and any move line reatined
+            is incomplete.
         """
         Move = self.env["stock.move"]
         self.ensure_one()
 
         mls = self.move_line_ids
 
         # Raise an error if any of the move lines are partially fulfilled
-        if mls and any(ml.qty_done != 0 and ml.qty_done < ml.product_uom_qty for ml in mls):
-            raise ValidationError(
-                _("You cannot create a backorder for %s with a move line qty less than expected!")
-                % self.picking_id.name
-            )
+        self._assert_move_lines_not_partially_fulfilled()
 
         # Check if there is something to split
-        if self.quantity_done == 0:
+        #   * If the quantity_done on the move is zero, and no move lines are passed,
+        #   there is no need to split up the move.
+        #   * If the mls_to_keep cover the whole move
+        # We just return self, but with it kicked out of the picking.
+        if (self.quantity_done == 0 and not mls_to_keep) or (
+            mls_to_keep and sum(mls_to_keep.mapped("product_uom_qty")) == self.product_uom_qty
+        ):
             # Clear the picking information
             self.write({"picking_id": False})
             if mls:
                 mls.write({"picking_id": False})
             return self
-        elif self.product_uom_qty == self.quantity_done:
+
+        # Get the mls_to_keep when not specified in the call
+        if mls_to_keep is None:
+            mls_to_keep = mls.get_lines_done()
+
+        mls_to_move = mls - mls_to_keep
+        # Ensure that if any mls_to_move are done, all mls_to_keep are as well!
+        if mls_to_move.get_lines_done() and mls_to_keep.get_lines_incomplete():
+            # TODO: What would be a good error message here?
+            raise ValidationError(
+                _(
+                    "You cannot move done move lines into a new move, if the existing move has incomplete lines"
+                )
+            )
+
+        # TODO: Over pick on a move line? How to handle this with custom configurations.
```

Comment:
Maybe not as important but can we add this as a "WARNING" in the docstring? I'd rather it be very clear in the docstring when someone reads about the function.